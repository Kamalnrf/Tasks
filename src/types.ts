type Task = {
  id: string
  title: string
  startTime: string | null
  endTime: string | null
  tags: Array<Tag>
}

type Tag = {
  name: string
  id: string
}

export {Task, Tag}
