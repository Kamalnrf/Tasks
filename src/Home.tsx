import React, {useReducer} from 'react'
import {FlatList, StyleSheet, Text, View} from 'react-native'
import {useSafeAreaInsets} from 'react-native-safe-area-context'
import {colors} from './constants'
import ActiveTask from './components/ActiveTask'
import NewTask from './components/NewTask'
import Button from './components/Button'
import {useTasks} from './hooks/useTasks'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.black_212227,
    paddingHorizontal: 10,
    justifyContent: 'space-between',
  },
  title: {
    fontFamily: 'Roboto_700Bold',
    lineHeight: 94,
    fontSize: 30,
    color: colors.white,
  },
  footer: {
    padding: 5,
    marginHorizontal: -10,
    backgroundColor: colors.black_2C2D2F,
  },
  footerText: {
    fontFamily: 'Roboto_400Regular',
    lineHeight: 46,
    fontSize: 15,
    color: colors.blue_92A9EC,
    alignSelf: 'flex-end',
    paddingRight: 10,
  },
})

/*
 TODO
 [ ] - Add E2E Tests
 [ ] - Add Tasks Search
 [ ] - Add Date Range Filter
*/

type State = {
  editingTaskId: string | undefined
  createNewTask: boolean
}

const initialState: State = {
  editingTaskId: undefined,
  createNewTask: false,
}

const Home = () => {
  const {top, bottom} = useSafeAreaInsets()
  const [state, setState] = useReducer(
    (s: State, a: Partial<State>) => ({...s, ...a}),
    initialState,
  )
  const {data} = useTasks()

  return (
    <>
      <View style={[styles.container, {paddingTop: top}]}>
        <Text style={styles.title}>Tasks</Text>
        <FlatList
          data={data?.tasks}
          keyExtractor={(item) => String(item.id)}
          renderItem={({item}) => (
            <ActiveTask
              {...item}
              onEdit={() =>
                setState({
                  editingTaskId: item.id,
                })
              }
            />
          )}
          ItemSeparatorComponent={() => <View style={{height: 20}} />}
          showsVerticalScrollIndicator={false}
          ListEmptyComponent={() => (
            <Button
              onPress={() => {
                setState({
                  createNewTask: true,
                })
              }}
              textStyle={{
                alignSelf: 'center',
                fontSize: 20,
              }}
            >
              Create new task +
            </Button>
          )}
        />
        <View style={[styles.footer, {paddingBottom: bottom}]}>
          <Button
            onPress={() => {
              setState({
                createNewTask: true,
              })
            }}
          >
            Create Task
          </Button>
        </View>
      </View>
      {!!state.editingTaskId || state.createNewTask ? (
        <NewTask
          task={data?.tasks?.find((task) => task.id === state.editingTaskId)}
          isEditing={!!state.editingTaskId}
          onClose={() =>
            setState({
              editingTaskId: undefined,
              createNewTask: false,
            })
          }
        />
      ) : (
        <></>
      )}
    </>
  )
}

export default Home
