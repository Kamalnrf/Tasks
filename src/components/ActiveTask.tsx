import React from 'react'
import {FlatList, StyleSheet, Text, View} from 'react-native'
import {colors} from '../constants'
import {Task} from '../types'
import Tag from './Tag'
import TimeProgress from './TimeProgreess'

const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.black_2C2D2F,
    borderRadius: 5,
    padding: 15,
    flexDirection: 'row',
  },
  taskDetailsContainer: {
    flex: 1,
    marginLeft: 20,
  },
  buttonsContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    flexWrap: 'wrap',
  },
  buttonTxt: {
    fontFamily: 'Roboto_400Regular',
    fontSize: 15,
  },
  titleTxt: {
    color: colors.white,
    fontFamily: 'Roboto_500Medium',
    fontSize: 20,
    lineHeight: 30,
  },
  buttonTxtLink: {
    color: colors.blue_92A9EC,
  },
})

const _isTaskRunning = (endTime: string | null, startTime: string | null) => {
  if (startTime && !endTime) return true
  else if (Date.parse(endTime as string) > Date.parse(startTime as string))
    return true
  return false
}

const ActiveTask = ({title, startTime, tags, endTime}: Task) => {
  const isTaskRunning = _isTaskRunning(endTime, startTime)

  return (
    <View style={styles.container}>
      <TimeProgress
        isTaskRunning={isTaskRunning}
        startTime={isTaskRunning ? new Date(startTime as string) : null}
      />
      <View style={styles.taskDetailsContainer}>
        <Text style={styles.titleTxt}>{title}</Text>
        <FlatList
          data={tags}
          keyExtractor={(item) => String(item.id)}
          renderItem={({item}) => <Tag key={item.id} name={item.name} />}
          horizontal={true}
          showsHorizontalScrollIndicator={false}
          style={{
            marginRight: -20,
          }}
        />
        <View style={styles.buttonsContainer}>
          {isTaskRunning ? (
            <Text
              style={[
                styles.buttonTxt,
                styles.buttonTxtLink,
                {paddingRight: 10},
              ]}
            >
              Stop Task
            </Text>
          ) : (
            <Text
              style={[
                styles.buttonTxt,
                styles.buttonTxtLink,
                {paddingRight: 10},
              ]}
            >
              Mark Completed
            </Text>
          )}
          <Text style={[styles.buttonTxt, styles.buttonTxtLink]}>
            Edit Task
          </Text>
        </View>
      </View>
    </View>
  )
}

export default ActiveTask
