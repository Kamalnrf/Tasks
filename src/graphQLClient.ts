import {GraphQLClient} from 'graphql-request'

const endpoint = 'https://test-323.herokuapp.com/v1/graphql'
const graphQLClient = new GraphQLClient(endpoint, {
  headers: {
    // ðŸ˜±
    authorization:
      'Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Im1xTFFXMDlNTUxRMUNNZGJpV3cwSyJ9.eyJodHRwczovL2hhc3VyYS5pby9qd3QvY2xhaW1zIjp7IngtaGFzdXJhLWRlZmF1bHQtcm9sZSI6InVzZXIiLCJ4LWhhc3VyYS1hbGxvd2VkLXJvbGVzIjpbInVzZXIiXSwieC1oYXN1cmEtdXNlci1pZCI6Imdvb2dsZS1vYXV0aDJ8MTA0NjY5NTI5OTM2OTYzOTk3MTIxIn0sImdpdmVuX25hbWUiOiJLYW1hbCIsImZhbWlseV9uYW1lIjoiTXVra2FtYWxhIiwibmlja25hbWUiOiJrYW1hbG5yZiIsIm5hbWUiOiJLYW1hbCBNdWtrYW1hbGEiLCJwaWN0dXJlIjoiaHR0cHM6Ly9saDMuZ29vZ2xldXNlcmNvbnRlbnQuY29tL2EtL0FPaDE0R2dEWkppTGk4UGJjMjY5MVpRcG55TmRmcUhyZmlwMFEyZEVXNkZicGJNPXM5Ni1jIiwibG9jYWxlIjoiZW4tR0IiLCJ1cGRhdGVkX2F0IjoiMjAyMS0wMi0yNFQxMDo0NTozNi41MTZaIiwiaXNzIjoiaHR0cHM6Ly90ZXN0LTMyMy51cy5hdXRoMC5jb20vIiwic3ViIjoiZ29vZ2xlLW9hdXRoMnwxMDQ2Njk1Mjk5MzY5NjM5OTcxMjEiLCJhdWQiOiJNclVTM3NZTEpUU1paMzJpUjN4OUhwQWJ3MzlWVVJVaCIsImlhdCI6MTYxNDE2MzUzNiwiZXhwIjoxNjE0MTk5NTM2LCJhdF9oYXNoIjoiUGMxYXFfWTN3eFRabkV2eGR5SkthQSIsIm5vbmNlIjoiSUptbnFVeGtZOGlRS2tYUH41R35KXzN-Z08uT2Z5TUkifQ.1kVrnU-q-1snOIV-y3pdIZ2grqOGsVQ2mlWsfg4d8NnsELkcUTizETZCInSwXoLO0c119wPRNoHL2F3Rj0FKdiJposhAO-_E4ShaCuUe4cMp2jpHH5dixHd79xCp-JXnpEgF0-_dnMCZ7Zvxr8Wl35-Dsfu3CmcCPlpjctVXp1g2jTeaQRQqUEt39ucBlWHuhqtLgMt_zM-0q12hZAEGBanJ2pCcsnuW-ripOXkaE83d4N4bIKNJFGrAjHJxmhx7mfyxnctm_wsC7TqJoQG38fQy9VLVf8gT0LgDg2EFYHKzlafy2q4g1aF_3u6pSbst8TEMak7v1hwf-BJUBUB18Q',
  },
})

export default graphQLClient
