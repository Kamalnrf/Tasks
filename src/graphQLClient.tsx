import {GraphQLClient} from 'graphql-request'

const endpoint = 'https://test-323.herokuapp.com/v1/graphql'
const graphQLClient = new GraphQLClient(endpoint, {
  headers: {
    // ðŸ˜±
    authorization:
      'Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Im1xTFFXMDlNTUxRMUNNZGJpV3cwSyJ9.eyJodHRwczovL2hhc3VyYS5pby9qd3QvY2xhaW1zIjp7IngtaGFzdXJhLWRlZmF1bHQtcm9sZSI6InVzZXIiLCJ4LWhhc3VyYS1hbGxvd2VkLXJvbGVzIjpbInVzZXIiXSwieC1oYXN1cmEtdXNlci1pZCI6Imdvb2dsZS1vYXV0aDJ8MTA0NjY5NTI5OTM2OTYzOTk3MTIxIn0sImdpdmVuX25hbWUiOiJLYW1hbCIsImZhbWlseV9uYW1lIjoiTXVra2FtYWxhIiwibmlja25hbWUiOiJrYW1hbG5yZiIsIm5hbWUiOiJLYW1hbCBNdWtrYW1hbGEiLCJwaWN0dXJlIjoiaHR0cHM6Ly9saDMuZ29vZ2xldXNlcmNvbnRlbnQuY29tL2EtL0FPaDE0R2dEWkppTGk4UGJjMjY5MVpRcG55TmRmcUhyZmlwMFEyZEVXNkZicGJNPXM5Ni1jIiwibG9jYWxlIjoiZW4tR0IiLCJ1cGRhdGVkX2F0IjoiMjAyMS0wMi0yM1QyMDo0MzozMC41NjNaIiwiaXNzIjoiaHR0cHM6Ly90ZXN0LTMyMy51cy5hdXRoMC5jb20vIiwic3ViIjoiZ29vZ2xlLW9hdXRoMnwxMDQ2Njk1Mjk5MzY5NjM5OTcxMjEiLCJhdWQiOiJNclVTM3NZTEpUU1paMzJpUjN4OUhwQWJ3MzlWVVJVaCIsImlhdCI6MTYxNDExMzAxMSwiZXhwIjoxNjE0MTQ5MDExLCJhdF9oYXNoIjoibnVqSnd0RVVSYzRCdFBiLWJxRmRlQSIsIm5vbmNlIjoiRW16TUk1ZnJwMGxKNX45b345Vm5JcnpZeFlscVg0aS0ifQ.aMQQAn_y8Wl2AzDfulVc-HG559zZJDcujaUiOupPfCeJ9z9qJEmNyuGmyR2s9PiJTtXsVL2zUfM7pzszLaabV5bKW_n1lp5FzAGKHpwgFySJCX981RhIcNas5eVtlPT6BfbC4qXGyTaym-0L1AIJdHhPHpTMJyV528ryPpi6iavEZoW2lF6xp3SEQedertFj6ILeOgllKeKIASgQlBQk74urlZ5pjI0NtftypFB8xD9LaeUlOmjoeVz1lVpIrVc-f7DpK1yArSXT88EU6Tzyk-0oGPEVPJOzzAx5uS8PGK6FZi9GApI025FUFfeAsSrfedxKtRoI_tQd86cE0UpLbw',
  },
})

export default graphQLClient
