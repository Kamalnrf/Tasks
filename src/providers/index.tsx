import React from 'react'
import {
  SafeAreaProvider,
  initialWindowMetrics,
} from 'react-native-safe-area-context'
import {QueryClient, QueryClientProvider} from 'react-query'

const queryClient = new QueryClient()

const AppProvider = ({children}: {children: React.ReactNode}) => {
  return (
    <QueryClientProvider client={queryClient}>
      <SafeAreaProvider initialMetrics={initialWindowMetrics}>
        {children}
      </SafeAreaProvider>
    </QueryClientProvider>
  )
}

export default AppProvider
