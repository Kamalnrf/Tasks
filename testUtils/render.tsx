import React from 'react'
import {render as rtlRender} from '@testing-library/react-native'
import {QueryClient, QueryClientProvider} from 'react-query'
import {SafeAreaProvider} from 'react-native-safe-area-context'

const queryClient = new QueryClient()

function render(ui: JSX.Element, {...renderOptions} = {}) {
  function Wrapper({children}: {children: JSX.Element}) {
    return (
      <QueryClientProvider client={queryClient}>
        <SafeAreaProvider
          initialMetrics={{
            frame: {x: 0, y: 0, width: 0, height: 0},
            insets: {top: 0, left: 0, right: 0, bottom: 0},
          }}
        >
          {children}
        </SafeAreaProvider>
      </QueryClientProvider>
    )
  }
  return rtlRender(ui, {wrapper: Wrapper, ...renderOptions})
}

export * from '@testing-library/react-native'

export {render}
