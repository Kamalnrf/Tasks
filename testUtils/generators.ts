import faker from 'faker'
import {Tag, Task} from '../src/types'

function generateList<T>(fn: (override?: {}) => T) {
  return (n: number): Array<T> =>
    new Array(n).fill('').reduce((acc, _) => [...acc, fn()])
}

export const generateTag = (override = {}): Tag => ({
  id: faker.random.uuid(),
  name: faker.name.title(),
  ...override,
})

export const generateTags = generateList<Tag>(generateTag)

export const generateTask = (override = {}): Task => ({
  id: faker.random.uuid(),
  endTime: faker.date.recent().toISOString(),
  startTime: faker.date.past().toISOString(),
  tags: generateTags(10),
  title: faker.name.title(),
  ...override,
})
