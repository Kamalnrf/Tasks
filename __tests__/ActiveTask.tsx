import React from 'react'
import ActiveTask from '../src/components/ActiveTask'
import {Task} from '../src/types'
import {generateTask} from '../testUtils/generators'
import {fireEvent, render} from '../testUtils/render'

const props = (override?: Partial<Task>) => ({
  onEdit: jest.fn(),
  ...generateTask(),
  ...override,
})

describe('Test Active Task', () => {
  it('should render active task which is not running', () => {
    const _props = props({
      startTime: null,
      endTime: null,
    })
    const {queryByText} = render(<ActiveTask {..._props} />)

    expect(queryByText('Start Task')).not.toBeNull()
    expect(queryByText('Edit Task')).not.toBeNull()
    expect(queryByText(_props.title)).not.toBeNull()
  })
  it('should render active task which is running', () => {
    const _props = props({
      startTime: new Date().toISOString(),
      endTime: null,
    })

    const {queryByText} = render(<ActiveTask {..._props} />)

    expect(queryByText('Mark Completed')).not.toBeNull()
    expect(queryByText('Edit Task')).not.toBeNull()
    expect(queryByText(_props.title)).not.toBeNull()
  })

  it('should trigger onEdit on tapping edit task', () => {
    const _props = props({
      startTime: null,
      endTime: null,
      title: 'New Task',
      tags: [],
      id: '1',
    })
    const {getByText, toJSON} = render(<ActiveTask {..._props} />)

    fireEvent.press(getByText('Edit Task'))
    expect(_props.onEdit).toHaveBeenCalledWith()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should render all tags of the task', () => {
    const _props = props()
    const {getByText} = render(<ActiveTask {..._props} />)

    _props.tags.map((tag) => {
      expect(getByText(tag.name)).not.toBeNull()
    })
  })
})
