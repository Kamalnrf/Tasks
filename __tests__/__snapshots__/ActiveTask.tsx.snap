// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Active Task should trigger onEdit on tapping edit task 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#2C2D2F",
        "borderRadius": 5,
        "flexDirection": "row",
        "padding": 15,
      }
    }
  >
    <RNSVGSvgView
      accessibilityRole="progressbar"
      bbHeight={100}
      bbWidth={100}
      focusable={false}
      height={100}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "transform": Array [
              Object {
                "rotateZ": "270deg",
              },
            ],
          },
          Object {
            "flex": 0,
            "height": 100,
            "width": 100,
          },
        ]
      }
      width={100}
    >
      <RNSVGGroup
        matrix={
          Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ]
        }
      >
        <RNSVGDefs>
          <RNSVGLinearGradient
            gradient={
              Array [
                0,
                -16752474,
                1,
                -16734567,
              ]
            }
            gradientTransform={null}
            gradientUnits={0}
            name="grad1"
            x1="0%"
            x2="100%"
            y1="0%"
            y2="0%"
          />
          <RNSVGClipPath
            name="clip"
          >
            <RNSVGCircle
              cx="50%"
              cy="50%"
              r={41}
            />
          </RNSVGClipPath>
        </RNSVGDefs>
        <RNSVGCircle
          cx={50}
          cy={50}
          propList={
            Array [
              "stroke",
              "strokeWidth",
            ]
          }
          r={45}
          stroke={4288454050}
          strokeWidth="10"
        />
        <RNSVGCircle
          cx={50}
          cy={50}
          propList={
            Array [
              "stroke",
              "strokeWidth",
              "strokeDasharray",
              "strokeDashoffset",
              "strokeLinecap",
            ]
          }
          r={45}
          stroke={
            Array [
              1,
              "grad1",
            ]
          }
          strokeDasharray={
            Array [
              "282.7433388230814",
              "282.7433388230814",
            ]
          }
          strokeDashoffset={0}
          strokeLinecap={1}
          strokeWidth="10"
        />
        <RNSVGLine
          matrix={
            Array [
              1,
              0,
              -0,
              1,
              50,
              0,
            ]
          }
          propList={
            Array [
              "stroke",
              "strokeWidth",
            ]
          }
          stroke={4288454050}
          strokeWidth="7"
          x1={0}
          x2={0}
          y1={25}
          y2={75}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontFamily": "Roboto_500Medium",
            "fontSize": 20,
            "lineHeight": 30,
          }
        }
      >
        New Task
      </Text>
      <RCTScrollView
        data={Array []}
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={true}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        showsHorizontalScrollIndicator={false}
        stickyHeaderIndices={Array []}
        style={
          Object {
            "marginRight": -20,
          }
        }
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "flex-end",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "flex-end",
                  "color": "#92A9EC",
                  "fontFamily": "Roboto_400Regular",
                  "fontSize": 15,
                  "lineHeight": 46,
                  "paddingRight": 10,
                },
                undefined,
              ]
            }
          >
            Start Task
          </Text>
        </View>
        <View
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "flex-end",
                  "color": "#92A9EC",
                  "fontFamily": "Roboto_400Regular",
                  "fontSize": 15,
                  "lineHeight": 46,
                  "paddingRight": 10,
                },
                undefined,
              ]
            }
          >
            Edit Task
          </Text>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
